feat: (#162) 회원 탈퇴 기능 추가

jungyoonho128 opened the pull request:
관련 이슈
- #162 

📝 제목
- 회원 탈퇴 추가 및 매니저 역할 보유 여부 확인
<!-- [작업 유형] 작업 내용 요약 (ex: [Feature] 회원가입 API 개발) -->

📋 작업 내용
- [ ] 회원 탈퇴 구현
- [ ] 탈퇴 전 매니저 역할 보유 여부 확인 및 예외처리
- [ ] Swagger 작성

🔧 기술 변경
- findManagedGroupsd으로 그룹 매니저 확인
 <!-- 새로운 모듈/패키지 추가 -->
 <!-- 폴더/파일 구조 변경 -->

🚀 테스트 사항(선택)
- 기능 테스트 완료
 <!--기능 테스트 완료 -->
 <!--API(Postman/Swagger) 테스트 -->
 <!-- 예외 처리 확인 -->

## 💬 리뷰 요구사항 (선택)  
<!-- 리뷰 요청 사항을 여기에 적어주세요 (선택사항) -->

jungyoonho128 committed:
feat: (#162) 회원 탈퇴 기능 추가

rrgks6221 commented on src/auth/auth.service.ts:
```diff
@@ -84,6 +84,10 @@ export class AuthService {
 
+  async deleteAccount(userId: number): Promise<void> {
+    await this.usersService.deleteUser(userId);
+  }
```
안쓰는 코드

```suggestion
  async deleteAccount(userId: number): Promise<void> {
    await this.usersService.deleteUser(userId);
  }
```

rrgks6221 commented on src/users/users.service.ts:
```diff
@@ -335,4 +341,43 @@ export class UsersService {
+    const replacementImageKey = this.getRandomDefaultImageKey();
+
+    await this.usersRepository.purgeUserDataExceptContent(userId, {
+      username: anonymizedUsername,
```
// 트렌젝션 처리는 알아서
    await this.usersRepository.updateUser(userId, {
      username: null,
      name: '탈퇴한 사용자',
      email: null,
      emailVerified: false,
      password: null,
      profileImgPath: replacementImageKey,
      nameChangedAt: null,
    });
    // userRepository에서 userGroup과 oAUthAccount를 관리하기 때문에 repository를 접근하여 사용해도 됨
    // await this.usersRepository.deleteManyUserGroup
    // await this.usersRepository.oAuthAccount
    // userNotification은 user에 종속된 하위 도메인 및 데이터이기 때문에 repository를 접근해 됨
    // await this.userNotification.deleteMany();
    // postLike같은 경우에는 Post라는 user와 동등한 최상위 도메인의 하위 도메인이기 때문에 repository에 접근하지 않고 postSErvice에서 기능을 제공하고 가져다 써야함
    // 다만 read findOne, findALl같은 읽기에 대해서는 repository를 가져와 쓰는 경우도 있음
    // service로 제공하야하는 이유는 만약 postLike가 삭제됐을 때 후속 작업이 필요하다면 그러한 로직이 userService에 추가가 되고
    // 응집성이 떨어지기 때문에 postLikeService에서 적절한 처리를 해주고 user도메인에서는 그러한 파생되는 동작을 몰라야함
    // await this.postLikeRepository.deleteMany();

rrgks6221 commented on src/users/users.service.ts:
```diff
@@ -335,4 +341,43 @@ export class UsersService {
+            err,
+          ),
+        );
+    }
```
사용할 변수는 로직과 가까운곳에 위치

const hasCustomImage = Boolean(user.profileImgPath);
    const isDefaultImage = user.profileImgPath
      ? this.defaultImageKeys.includes(user.profileImgPath)
      : false;

```suggestion
    if (hasCustomImage && !isDefaultImage && user.profileImgPath) {
      await this.s3Service
        .deleteFile(user.profileImgPath)
        .catch((err) =>
          console.error(
            `사용자(${userId}) 탈퇴 시 프로필 이미지 삭제 실패: ${user.profileImgPath}`,
            err,
          ),
        );
    }
```

rrgks6221 commented on src/users/users.service.ts:
```diff
@@ -182,6 +183,11 @@ export class UsersService {
+  private generateDeletedUsername(userId: number): string {
+    const suffix = Date.now().toString(36).slice(-4);
+    return `del${userId}${suffix}`.slice(0, 20);
+  }
```
username 제거해도 되기 때문에 삭제해도 됨
```suggestion
```

jungyoonho128 committed:
refactor: (#162) 사용자 탈퇴 기능 개선 및 deletedAt 필드 추가

- User 모델에 deletedAt 필드 추가 및 Prisma 마이그레이션 적용
- 사용자 탈퇴 시 개인정보 익명화 및 deletedAt 소프트 삭제 방식으로 변경
- UsersRepository 조회 메서드에서 deletedAt 필드 기반 필터링 적용
- 그룹 매니저 권한 보유 시 탈퇴 제한 로직 추가
- S3 프로필 이미지 삭제 시 Nest Logger로 에러 기록 방식 개선
- username 등 필수 필드 익명값으로 덮어쓰기 유지

jungyoonho128 committed:
refactor: (#162) 사용자 매니저 권한 확인 기능 추가

- 그룹 내 마지막 매니저인 경우 탈퇴 불가 로직 추가
- UsersRepository에 findGroupsWhereUserIsSoleManager 메서드 구현

jungyoonho128 committed:
refactor: (#162) 사용자 탈퇴 기능 추가 및 이벤트 발행 구현

- 사용자 탈퇴 로직을 UserDeletionService로 분리하여 책임 분리
- 탈퇴 후 S3 이미지 삭제를 이벤트 기반으로 비동기 처리
- UserDeletionListener 추가로 이벤트 리스너 구현
- UsersModule에 EventEmitterModule 및 UserDeletionListener 주입

jungyoonho128 committed:
refactor: (#162) 사용자 탈퇴 관련 기능 개선 및 이벤트 리스너 제거

- 사용자 탈퇴 로직을 UserDeletionService로 집중하여 책임 분리
- S3 프로필 이미지 삭제 등 후처리 로직을 서비스 내부에서 직접 처리하도록 변경
- 이벤트 기반 처리 및 UserDeletionListener 제거

jungyoonho128 committed:
chore: (#162) trigger CI

jungyoonho128 committed:
refactor: (#162) 사용자 탈퇴 시 프로필 이미지 S3 직접 삭제하도록 로직 변경

- 사용자 탈퇴 시 usersService.resetProfileImage 호출 대신 S3에서 직접 이미지 삭제하도록 변경
- DB의 profileImgPath는 이미 purgeUserDataExceptContent에서 기본 이미지로 세팅되므로, S3 삭제만 별도로 처리하여 코드 간결성 향상
